### PR TITLE
Make salinity data in the Information tab editable.

### PR DESCRIPTION
**Pull request by willemferguson**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Changes made:

This series of commits make the salinity data editable. The combobox for selecting water type (thus salinity) is activated by a preference setting in the Preferences->General screen. If the preference is not checked the behaviour of the Information screen remains exactly as it is at present.

There are two situations where a diver would like to edit the salinity. a) when a dive is entered manually. b) when the dc has not been set to the appropriate water type. The latter case could produce unintended consequences for the dive profile: if the water salinity (=density) is changed after the dive, the depth profile and ceiling information are also changed and the ceiling data may not reflect the dive accurately any more. To address this issue and others, any user-supplied salinity is stored in the dive structure in a variable that is not used for any dive calculations. It is really just a note for the diver to indicate the true water type that was dived and does not change any of the salinity values used for decompression calculations. If a salinity value is chosen (by using the combobox in the Information tab) to replace the displayed salinity, an exclamation mark is shown next to the water type to indicate that the selected water type and salinity do not correspond to the salinity provided by the dc. The indicated value for salinity can be reset to that of the dc by selecting "use dc" option of the combobox. The exclamation mark is not shown for manually-entered dives. Obviously, in the case of manually-entered dives, the option of "use dc" does not provide any useful information. This facility allows saving salinity data in a way that do not affect the logged dive, its depth profile or its dive ceiling calculations. 

These changes turned out, against expectations, to be quite invasive to the code, affecting 16 files. I have broken down the new code into 5 steps:

1. Create preference setting in Preferences->General screen. The preference is saved and reloaded upon restart of subsurface just like other preferences. Changes to 7 files.
2. Create the combobox for selecting water types and its required behaviour in the Information tab. One requirement is that the user may select the same water type as was displayed before opening the combobox: comboboxes normally do not allow this. Changes to 4 files.
3. Create the file I/O for storing and loading the salinity value selected using the combobox. Changes to four files.
4. Create the undo infrastructure for changes in salinity. Changes to 4 files.
5. A small amount of cleanup at the end of the changes. Changes to 2 files.

### Mentions:

@dirkhh @atdotde @bstoeger @glance- @sfuchs79

### CODE REVIEW

File: desktop-widgets/tab-widgets/TabDiveInformation.ui
Comment (dirkhh) [inline desktop-widgets/tab-widgets/TabDiveInformation.ui]: why is the text now `!` ?

File: desktop-widgets/tab-widgets/TabDiveInformation.ui
Comment (dirkhh) [inline desktop-widgets/tab-widgets/TabDiveInformation.ui:457]: hand editing these files is not for the faint of heart  - and rarely a good idea. Can you explain what you tried to achieve here?

File: core/load-git.c
Comment (dirkhh) [inline core/load-git.c]: please look at the comment in line 989 (new line numbers) above., These need to be sorted or the code won't work.

**Review by dirkhh — CHANGES_REQUESTED** [review]
Thanks for breaking it down into smaller commits - but those naturally are ordered, so adding the "step 1" "step 2" etc is not required and doesn't match our usual style.
I noticed a couple of issues and pointed them out in comments.
I'll let @bstoeger comment on the undo code ;-)

File: core/load-git.c
Comment (willemferguson) [reply core/load-git.c]: I will correct this. There is actually a comment in the code that the items need to be in sorted order. I will wait for the comments from others so that I can consolidate them into a single commit. I also need to add translation of strings and remove unwanted #includes. As is at the moment, visibility does not work.

Comment (willemferguson) [reply core/load-git.c]: Done

File: desktop-widgets/tab-widgets/TabDiveInformation.ui
Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.ui:457]: Done

File: desktop-widgets/tab-widgets/TabDiveInformation.ui
Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.ui]: I hope this is solved. This is not a bug, it is meant to be like that.

**Review by dirkhh — COMMENTED** [review]
git would allow you to actually make the changes to your original commits instead of having following commits clean up things that went sideways in earlier ones.
I realize that this is a bit like open heart surgery and maybe it's not for everyone. If you want I can restructure things a bit when merging this by hand.
Since you are still making changes to the "features" of this PR (like adding the icon for overwriting salinity), do you consider this PR "ready to be merged" or are you still working on it?

File: desktop-widgets/tab-widgets/TabDiveInformation.cpp
Comment (dirkhh) [inline desktop-widgets/tab-widgets/TabDiveInformation.cpp]: I think this deserves a comment why the different code for Windows is necessary

Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.cpp]: This commit with respect to the overwrite indication is ready for merging. 

The OS-specific CSS styling is an attempt to get the Information tab to look fairly similar in the three platforms we support. With the above code the Win rendering is:
Win
And the Linux rendering is:
Linux
which are roughly comparable. With the previous code the Mac rendering was very unfriendly but I have no idea what it looks like on a Mac now. Assuming that Qt on a Mac and on Linux are fairly similar (both Unix-derived) I hope the Mac rendering will also be much more friendly now

Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.cpp]: Of course I will do more changes after @glance and @bstoeger have inspected the code.

File: commands/command_edit.cpp
Comment (bstoeger) [inline commands/command_edit.cpp]: Copy & paste? Should probably read "Water type".

File: desktop-widgets/tab-widgets/TabDiveInformation.cpp
Comment (bstoeger) [inline desktop-widgets/tab-widgets/TabDiveInformation.cpp]: Coding style: superfluous empty line.

File: desktop-widgets/tab-widgets/TabDiveInformation.h
Comment (bstoeger) [inline desktop-widgets/tab-widgets/TabDiveInformation.h:30]: I'm not convinced that we need this member variable: It is only queried in one place. Can't we simply do the check there?

**Review by bstoeger — COMMENTED** [review]
Mostly a few coding style comment.

What I am a bit confused about is why the edit command says "Combo" but sets a density. `editWaterTypeUser` would seem more logical to me.

File: commands/command_edit.cpp
Comment (willemferguson) [reply commands/command_edit.cpp]: Done

File: commands/command.cpp
Comment (willemferguson) [reply commands/command.cpp]: Done

File: commands/command_edit.cpp
Comment (willemferguson) [reply commands/command_edit.cpp:275]: Done

File: commands/command_edit.cpp
Comment (willemferguson) [reply commands/command_edit.cpp]: fixed

File: core/dive.h
Comment (willemferguson) [reply core/dive.h]: fixed

File: desktop-widgets/tab-widgets/TabDiveInformation.cpp
Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.cpp]: fixed

File: desktop-widgets/tab-widgets/TabDiveInformation.cpp
Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.cpp]: fixed. Created enum.

File: desktop-widgets/tab-widgets/TabDiveInformation.cpp
Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.cpp]: fixed

File: desktop-widgets/tab-widgets/TabDiveInformation.cpp
Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.cpp]: fixed

File: desktop-widgets/tab-widgets/TabDiveInformation.cpp
Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.cpp]: fixed

File: desktop-widgets/tab-widgets/TabDiveInformation.h
Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.h:30]: I get a segmentation fault if I access dive->dc.model directly in the code. In general I have had quite a bit of problems of crashes when accessing current_dive's components.

**Comment by willemferguson** [comment]
In addition I added a tooltip to the exclamation mark icon. If users suddenly see this icon appearing they might not immediately appreciate its significance. The tooltip my help with that.

**Comment by dirkhh** [comment]
So what's the status of this (besides being in need of being rebased onto master)?
Should I do that + final review and merge?

**Comment by willemferguson** [comment]
Yes, I responded to all the comments that were offered. Currently it still says 'Waiting for 1 review". As far as I am concerned it is done.

**Review by dirkhh — APPROVED** [review]

**Comment by dirkhh** [comment]
> Yes, I responded to all the comments that were offered. Currently it still says 'Waiting for 1 review". As far as I am concerned it is done.

That was just me not having updated my review status.
I'll wait for the preferences UI to go in, then rebase and merge this.

**Comment by dirkhh** [comment]
I merged this with the latest master and the result has a couple of issues. My biggest concern is that when reading an existing dive file all dives appear to have "fresh water" set as dive type - which gets you a warning icon as the water type doesn't match the DC (all my dives are indeed salt water). So there's clearly something going wrong with the initialization in the absence of data when loading an existing file.

Also, I'm not quite sure what's going wrong with my Linux build - this looks fine on the Mac...
Screenshot_20191212_215447

I'll push my merged tree so you can look into this

**Pull request closed without merging**
